1.2 python3

cmazzullo opened the pull request:
Seeker 1.2, updated to work on Python3

- Ran 2to3 on the project, then munged the tests and settings a bit to get everything working (so far). 
- Changed the test settings to make it easier to get tests running and updated the docs to reflect this.

Commit:
Got a couple tests working to prep for Python 3 conversion

Commit:
Ran 2to3 on whole project

Commit:
Added test_mapping_refresh and test_termAggregate

Commit:
Updated the tests/settings.py file to make it possible to run tests

Also updated the docs/tests.rst file to reflect the changes

dcwatson commented:
I think making a python3-only version of this branch is a bit of a non-starter. Even a hypothetical  python3-only 1.3 branch feels strange, considering later versions support both python2 and python3. I'd say your options are updating this to support py2/3 (using six), or update your site to a later version of seeker.

Commit:
Added dependency on six + some related fixes
- Added urllib six checks in views.py
- Added python_2_unicode_compatible decorators in core/models.py and seeker/models.py
- Added six to requirements.txt

Commit:
PEP8 fixes (i.e. ran autopep8 on the whole codebase)

Commit:
Added two reindexing tests + added a six check to the reindex method

cmazzullo commented:
These commits update the branch to support Python 2 as well using six.

dcwatson commented on seeker/management/commands/reindex.py:
```diff
@@ -1,20 +1,27 @@
 try:
     from tqdm import tqdm as progress_iter
-except:
+except BaseException:
```
These should probably be `Exception` instead of `BaseException`.

dcwatson commented on seeker/management/commands/reindex.py:
```diff
@@ -32,49 +40,55 @@ def get_actions():
+    if six.PY2:
+        writer = io.BytesIO
+    elif six.PY3:
+        writer = io.StringIO
```
Why bytes vs. string for 2/3? Seems like this should be https://pythonhosted.org/six/#six.StringIO

dcwatson commented on seeker/mapping.py:
```diff
@@ -71,11 +73,11 @@ def to_elastic(self, value):
             return None
         elif isinstance(value, (list, tuple)):
-            return [unicode(v) for v in value]
+            return [str(v) for v in value]
```
`str` will behave differently between 2 and 3. You could use https://pythonhosted.org/six/#six.text_type or something like `django.utils.encoding.force_text`

dcwatson commented on seeker/query.py:
```diff
@@ -59,16 +61,17 @@ def instance(self):
         self.mapping = mapping
         self.query = query or {}
-        if isinstance(self.query, basestring):
+        if isinstance(self.query, str):
```
Should use https://pythonhosted.org/six/#six.string_types

dcwatson commented on seeker/views.py:
```diff
@@ -6,10 +6,17 @@
+
+import six
+
+if six.PY2:
```
Six has a "moves" module that lets you import from a single package for 2/3 without having to do explicit version checking like this:

https://pythonhosted.org/six/#module-six.moves

dcwatson reviewed (commented):
Some initial notes. Have you tried the unit tests on 2 and 3? I'm honestly not sure what kind of shape the tests are in for this branch, but if it's possible to get them running that would be good. Plus, given the changes in text/byte handling between Python versions, it may even be good to add some non-ASCII tests to make sure everything works.

Commit:
Replaced "catch BaseException" with "catch Exception"

Commit:
ISSUE 1 - Added empty tests/urls.py

Commit:
ISSUE 1 - Fixed up some six use
- used six.moves instead of explicit version checks
- switched from BytesIO to six.StringIO
- used six.text_type instead of str
- used six.string_types instead of str/basestring

Commit:
ISSUE 1 - PEP8

Commit:
ISSUE 1 - Changed filter test (it was trying to filter an analyzed field with raw text

cmazzullo commented:
Addressed those comments. It took some changes, but the tests are running, although coverage isn't great. All but one (a test for seeker.crossquery) are passing.

dcwatson commented on seeker/query.py:
```diff
@@ -77,15 +82,15 @@ def __init__(self, mapping, query=None, filters=None, facets=None, highlight=Non
         if isinstance(self.highlight, (list, tuple)):
             self.highlight = {'fields': {f: {'number_of_fragments': 0} for f in self.highlight}}
-        elif isinstance(self.highlight, basestring):
+        elif isinstance(self.highlight, str):
```
There a reason not to use `six.string_types` here (and below) as well?

dcwatson commented on seeker/query.py:
```diff
@@ -343,15 +359,15 @@ def __and__(self, other):
 
     def filter_spec(self, val):
-        if isinstance(val[1], (basestring, bool, int)):
+        if isinstance(val[1], (str, bool, int)):
```
This won't catch unicode on Py2 - use `six.string_types` in addition to bool/int (might have to concat tuples or do an `or`)

dcwatson commented on seeker/templatetags/seeker.py:
```diff
@@ -55,20 +60,23 @@ def string_format(value):
     if isinstance(value, datetime.date):
         return dateformat.format(value, settings.DATE_FORMAT)
-    return unicode(value)
+    return str(value)
```
`force_text` or `six.text_type`

dcwatson commented on seeker/templatetags/seeker.py:
```diff
@@ -124,20 +137,21 @@ def _highlight(obj, words):
         return values, was_highlighted
-    elif isinstance(obj, (unicode, str, int)):
-        s = escape(unicode(obj))
+    elif isinstance(obj, (str, int)):
```
See above.

dcwatson commented on seeker/utils.py:
```diff
@@ -67,7 +74,7 @@ def crossquery(query, suggest=None, limit=None, offset=None, hosts=None):
     es = Elasticsearch(hosts or getattr(settings, 'SEEKER_HOSTS', None))
     query = query or {}
-    if isinstance(query, basestring):
+    if isinstance(query, str):
```
`str` != `basestring`, use `six.string_types`

dcwatson commented on seeker/views.py:
```diff
@@ -274,8 +282,8 @@ def export(self, request):
             if isinstance(value, (list, tuple)):
-                value = '; '.join(unicode(v) for v in value)
-            return '"%s"' % unicode(value).replace('"', '""')
+                value = '; '.join(str(v) for v in value)
```
Another one.

Commit:
ISSUE 1 - More six.string_types, six.test_type use

Commit:
ISSUE 1 - Line endings, PEP8

cmazzullo commented:
Addressed those comments

dcwatson commented:
This is looking better to me. Have you tested this with your site yet? Once you do some preliminary testing to make sure there aren't any obvious bugs, I'll create a new 1.3 branch and merge this there. Then this will eventually be a 1.3 release on PyPI (seems like too big a change for 1.2.x).

cmazzullo commented:
I've done some preliminary testing on one of the dev servers (BioLINCC 3903) and haven't found any issues yet. We should have some more rigorous testing done in the near future.

ghost commented:
Dan, we've been testing these updates without any issues; I think we're good to go with a 1.3 branch now.

Chris, do you agree?

cmazzullo commented:
Yes, I think it's good to go

On Mon, Mar 4, 2019, 6:25 PM kirksnyder <notifications@github.com> wrote:

> Dan, we've been testing these updates without any issues; I think we're
> good to go with a 1.3 branch now.
>
> Chris, do you agree?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/imsweb/django-seeker/pull/68#issuecomment-469467636>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-H8zv2TQPhcuFoxzYzosRcVWFerTh1ks5vTatSgaJpZM4XkD6F>
> .
>
